src/D2L.CodeStyle.Analyzers/Immutability/StatelessFuncAnalyzer.cs

```diff
@@ -0,0 +1,153 @@
+
+namespace D2L.CodeStyle.Analyzers.Immutability {
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	public sealed class StatelessFuncAnalyzer : DiagnosticAnalyzer {
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+			Diagnostics.StatelessFuncIsnt
+		);
+
+		public override void Initialize( AnalysisContext context ) {
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction( RegisterAnalysis );
+		}
+
+		private void RegisterAnalysis( CompilationStartAnalysisContext context ) {
+
+			Compilation compilation = context.Compilation;
+
+			context.RegisterSyntaxNodeAction(
+				ctx => {
+					if( ctx.Node is ObjectCreationExpressionSyntax expr ) {
+						AnalyzeInvocation( ctx, expr );
+					}
+				},
+				SyntaxKind.ObjectCreationExpression
+			);
+		}
+
+		private void AnalyzeInvocation(
+			SyntaxNodeAnalysisContext context,
+			ObjectCreationExpressionSyntax syntax
+		) {
+
+			TypeSyntax type = syntax.Type;
+			ISymbol symbol = context
+				.SemanticModel
+				.GetSymbolInfo( type ).Symbol;
+
+			if( symbol == null ) {
+				return;
+			}
+
+			if( !IsStatelessFunc( symbol ) ) {
+				return;
+			}
+
+			ExpressionSyntax argument = syntax.ArgumentList.Arguments[ 0 ].Expression;
+			SyntaxKind kind = argument.Kind();
+			switch( kind ) {
+
+				// this is the case when a method reference is used
+				// eg Func<string, int> func = int.Parse
+				case SyntaxKind.SimpleMemberAccessExpression:
+					if( IsStaticMemberAccess( context, argument ) ) {
+						return;
+					}
+
+					// non-static member access means that state could
+					// be used / held.
+					// TODO: Look for [Immutable] on the member's type
+					// to determine if the non-static member is safe
+					break;
+
+				// this is the case when the left hand side of the
+				// lambda has parens
+				// eg () => 1, (x, y) => x + y
+				case SyntaxKind.ParenthesizedLambdaExpression:
+					if( !CheckForClosures( context, argument ) ) {
+						return;
+					}
+					break;
+
+				// this is the case when the left hand side of the
+				// lambda does not have parens
+				// eg x => x + 1
+				case SyntaxKind.SimpleLambdaExpression:
+					if( !CheckForClosures( context, argument ) ) {
+						return;
+					}
+					break;
+
+				default:
+					// should we do something else here because constructors
+					// of D2L.StatelessFunc<T> and friends should always take
+					// a func?
+					return;
+			}
+
+			var diag = Diagnostic.Create(
+				Diagnostics.StatelessFuncIsnt,
+				argument.GetLocation()
+			);
+
+			context.ReportDiagnostic( diag );
+		}
+
+		private static bool IsStaticMemberAccess(
+			SyntaxNodeAnalysisContext context,
+			ExpressionSyntax expression
+		) {
+			ISymbol symbol = context
+				.SemanticModel
+				.GetSymbolInfo( expression ).Symbol;
+
+			if( symbol == null ) {
+				return false;
+			}
+
+			return symbol.IsStatic;
+		}
+
+		private static bool CheckForClosures(
```

Comment:
nit: it might be more obvious what the return value is if this were named `HasCaptures` or `CapturesData` or something (vs. "Check" - it wasn't obvious to me which direction it was checking for)

Reply:
Fixed